msdsales/read_test.go

```diff
@@ -0,0 +1,292 @@
+			}
+		})
+	}
+}
+
+var contactsFirstPageResponse = `{
+		"@odata.context": "https://org5bd08fdd.api.crm.dynamics.com/api/data/v9.2/$metadata#contacts(fullname,emailaddress1,fax,familystatuscode)",
+		"@Microsoft.Dynamics.CRM.totalrecordcount": -1,
+		"@Microsoft.Dynamics.CRM.totalrecordcountlimitexceeded": false,
+		"@Microsoft.Dynamics.CRM.globalmetadataversion": "6012567",
+		"value": [
+		{
+		  "@odata.etag": "W/\"4372108\"",
+		  "fullname": "Heriberto Nathan",
+		  "emailaddress1": "heriberto@northwindtraders.com",
+		  "fax": "614-555-0122",
+		  "familystatuscode@OData.Community.Display.V1.FormattedValue": "Single",
+		  "familystatuscode": 1,
+		  "contactid": "cdcfa450-cb0c-ea11-a813-000d3a1b1223"
+		},
+		{
+		  "@odata.etag": "W/\"4372115\"",
+		  "fullname": "Dwayne Elijah",
+		  "emailaddress1": "dwayne@alpineskihouse.com",
+		  "fax": "281-555-0158",
+		  "familystatuscode@OData.Community.Display.V1.FormattedValue": "Single",
+		  "familystatuscode": 1,
+		  "contactid": "9fd4a450-cb0c-ea11-a813-000d3a1b1223"
+		}
+		],
+		"@odata.nextLink": "https://org5bd08fdd.api.crm.dynamics.com/api/data/v9.2/contacts?$select=fullname,emailaddress1,fax,familystatuscode&$skiptoken=%3Ccookie%20pagenumber=%222%22%20pagingcookie=%22%253ccookie%2520page%253d%25221%2522%253e%253ccontactid%2520last%253d%2522%257b9FD4A450-CB0C-EA11-A813-000D3A1B1223%257d%2522%2520first%253d%2522%257bCDCFA450-CB0C-EA11-A813-000D3A1B1223%257d%2522%2520%252f%253e%253c%252fcookie%253e%22%20istracking=%22False%22%20/%3E"
+	}`
+
+func Test_Read(t *testing.T) {
+	t.Parallel()
+
+	tests := []struct {
+		name         string
+		input        common.ReadParams
+		server       *httptest.Server
+		connector    Connector
+		expected     *common.ReadResult
+		expectedErrs []error
+	}{
+		{
+			name: "Mime response header expected",
+			server: httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+				w.WriteHeader(http.StatusTeapot)
+			})),
+			expectedErrs: []error{interpreter.MissingContentType},
+		},
+		{
+			name: "Correct error message is understood from JSON response",
+			server: httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+				w.Header().Set("Content-Type", "application/json")
+				w.WriteHeader(http.StatusBadRequest)
+				writeBody(w, `{
+					"error": {
+						"code": "0x80060888",
+						"message":"Resource not found for the segment 'conacs'."
+					}
+				}`)
+			})),
+			expectedErrs: []error{common.ErrBadRequest, errors.New("Resource not found for the segment 'conacs'")},
+		},
+		{
+			name: "Incorrect key in payload",
+			server: httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+				w.Header().Set("Content-Type", "application/json")
+				w.WriteHeader(http.StatusOK)
+				writeBody(w, `{
+					"garbage": {}
+				}`)
+			})),
+			expectedErrs: []error{errors.New("wrong request: wrong key 'value'")},
```

Comment:
Yes, this is a smell code. There was another test where I used type comparison. To be more general I added reflection check. For this test checking that error comes from `ajson` lib should be enough.
```
expectedErrs - used when you know exact error
expectedErrTypes - used less frequently and relies on type rather than content
```